### ISSUE TITLE
0.20 build fails on ppc64el

### ISSUE THREAD

**Issue by infinity0** [celluloid-player/celluloid#605]
https://buildd.debian.org/status/fetch.php?pkg=celluloid&arch=ppc64el&ver=0.20-1&stamp=1605564917&raw=0

~~~~
gcc -pthread -I/usr/include/gtk-3.0 -I/usr/include/at-spi2-atk/2.0 -I/usr/include/at-spi-2.0 -I/usr/include/dbus-1.0 -I/usr/lib/powerpc64le-linux-gnu/dbus-1.0/include -I/usr/include/gtk-3.0 -I/usr/include/gio-unix-2.0 -I/usr/include/cairo -I/usr/include/pango-1.0 -I/usr/include/fribidi -I/usr/include/harfbuzz -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/glib-2.0 -I/usr/lib/powerpc64le-linux-gnu/glib-2.0/include -DG_SETTINGS_ENABLE_BACKEND -DPACKAGE_LOCALEDIR=\""/usr/share/locale"\" -Impris -g -O2 -fdebug-prefix-map=/<<PKGBUILDDIR>>=. -fstack-protector-strong -Wformat -Werror=format-security -std=gnu99 -funsigned-char -fPIE -fPIC -Wall -Wextra -Wconversion -Winline -Wno-padded -Wno-unused-parameter -Wstrict-prototypes -Wmissing-prototypes -Werror=implicit-function-declaration -Werror=pointer-arith -Werror=init-self -Werror=format=2 -Werror=missing-include-dirs -Werror=date-time  -Wl,-z,relro -pie -Wl,-z,now -o celluloid celluloid-celluloid-main.o celluloid-celluloid-application.o celluloid-celluloid-common.o celluloid-celluloid-controller.o celluloid-celluloid-controller-actions.o celluloid-celluloid-controller-input.o celluloid-celluloid-control-box.o celluloid-celluloid-file-chooser.o celluloid-celluloid-header-bar.o celluloid-celluloid-main-window.o celluloid-celluloid-menu.o celluloid-celluloid-metadata-cache.o celluloid-celluloid-model.o celluloid-celluloid-mpv.o celluloid-celluloid-option-parser.o celluloid-celluloid-open-location-dialog.o celluloid-celluloid-player.o celluloid-celluloid-player-options.o celluloid-celluloid-playlist-widget.o celluloid-celluloid-plugins-manager.o celluloid-celluloid-plugins-manager-item.o celluloid-celluloid-preferences-dialog.o celluloid-celluloid-seek-bar.o celluloid-celluloid-shortcuts-window.o celluloid-celluloid-time-label.o celluloid-celluloid-video-area.o celluloid-celluloid-view.o celluloid-celluloid-mpv-wrapper.o mpris/celluloid-celluloid-mpris.o mpris/celluloid-celluloid-mpris-module.o mpris/celluloid-celluloid-mpris-base.o mpris/celluloid-celluloid-mpris-player.o mpris/celluloid-celluloid-mpris-track-list.o mpris/celluloid-celluloid-mpris-gdbus.o celluloid-celluloid-marshal.o media-keys/celluloid-celluloid-media-keys.o  -lgtk-3 -lgdk-3 -lpangocairo-1.0 -lpango-1.0 -lharfbuzz -latk-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lmpv -lepoxy -lm 
/usr/bin/ld: mpris/celluloid-celluloid-mpris-base.o: in function `celluloid_mpris_base_new':
./src/mpris/celluloid-mpris-base.c:374: undefined reference to `celluloid_mpris_org_mpris_media_player2_interface_info'
/usr/bin/ld: mpris/celluloid-celluloid-mpris-player.o: in function `celluloid_mpris_player_new':
./src/mpris/celluloid-mpris-player.c:855: undefined reference to `celluloid_mpris_org_mpris_media_player2_player_interface_info'
/usr/bin/ld: mpris/celluloid-celluloid-mpris-track-list.o: in function `celluloid_mpris_track_list_new':
./src/mpris/celluloid-mpris-track-list.c:590: undefined reference to `celluloid_mpris_org_mpris_media_player2_track_list_interface_info'
collect2: error: ld returned 1 exit status
~~~~

Perhaps you need to add some `#ifdefs` or something? The build succeeds on other architectures anyhow.

**Comment by gnome-mpv** [comment]
I tried to reproduce the issue in a debian (buster) ppc64el VM, but it builds fine there. I couldn't test on sid because mpv wouldn't install due to dependencies issues.

My guess would be that the gdbus code wasn't being generated properly for some reason. Is there some way to download `src/mpris/celluloid-mpris-gdbus.{c,h}` from the failing build? I think it might be helpful if we can compare them against the ones from successful builds.

It might be worth trying to build with meson as well. It has a wrapper for `gdbus-codegen` that might work better.

**Comment by infinity0** [comment]
I couldn't reproduce it either on a debian ppc64el machine running sid. However I ran the build again on the build daemon (which I don't have shell access to) and it failed with a different error, about a syntax error: https://buildd.debian.org/status/fetch.php?pkg=celluloid&arch=ppc64el&ver=0.20-1&stamp=1606264971&raw=0

I notice the build appears to be building `mpris/celluloid-mpris-gdbus.c` twice, is there a reason for this?

~~~~
/usr/bin/gdbus-codegen --c-namespace celluloid_mpris --generate-c-code mpris/celluloid-mpris-gdbus ../data/celluloid-mpris-gdbus.xml && \
/bin/sed -i '1i#pragma GCC diagnostic ignored \"-Wconversion\"' mpris/celluloid-mpris-gdbus.c
\
/usr/bin/gdbus-codegen --c-namespace celluloid_mpris --generate-c-code mpris/celluloid-mpris-gdbus ../data/celluloid-mpris-gdbus.xml && \
/bin/sed -i '1i#pragma GCC diagnostic ignored \"-Wconversion\"' mpris/celluloid-mpris-gdbus.c
\
~~~~

Perhaps these are being run in parallel and somehow this works fine on the other machines, but clobbers the file on this machine?

**Comment by infinity0** [comment]
I think you want to use `$(mpris_generated) &:` in `Makefile.am` instead of `:`, explanation here:

https://www.gnu.org/software/make/manual/html_node/Multiple-Targets.html

Using `:` means independent targets and so `make` indeed will run the recipe twice. I guess it's just luck that this machine triggered the bug more often than the other machines did. Maybe something to do with how `make` populates internal hashtables on that architecture or something.

**Comment by gnome-mpv** [comment]
That makes sense. I committed the fix in 76327750a417e9e2c5e1c68a6a2d31c0e67f7495.

Can you confirm that it fixes the problem?

**Comment by infinity0** [comment]
@gnome-mpv yes it works, thanks!
